File: sudo/src/main/java/com/roman/gurdan/sudo/game/cell/Cell.java
package com.roman.gurdan.sudo.game.cell;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Cell {

    public int row; // 行
    public int col; //列
    public int group; // 所在子宫格
    public int value = 0; // 值
    public boolean preSet = true; // 是否是预置值
    public boolean valid = true; // 是否正确
    public List<Integer> optional = new ArrayList<>();

    public Cell() {
    }

    public Cell(int row, int col, int group, int value) {
        this.row = row;
        this.col = col;
        this.group = group;
        this.value = value;
    }

    public void setGroup(int group) {
        this.group = group;
    }

    public void setValue(int value) {
        this.value = value;
    }

    public void setValid(boolean valid) {
        this.valid = valid;
    }



    public boolean fromJson(JSONObject jsonObject) throws JSONException {
        try{
            this.row = jsonObject.getInt("row");
            this.col = jsonObject.getInt("col");
            this.group = jsonObject.getInt("group");
            this.value = jsonObject.getInt("value");
            this.preSet = jsonObject.getBoolean("preSet");
            this.valid = jsonObject.getBoolean("valid");
        } catch (Exception e){
            throw e;
        }
        return true;
    }

    @Override
    public String toString() {
        return "{" +
                "row:" + row +
                ", col:" + col +
                ", group:" + group +
                ", value:" + value +
                ", preSet:" + preSet +
                ", valid:" + valid +
                '}';
    }

}

File: base/src/main/java/com/roman/garden/sudo/base/game/Cell.kt
package com.roman.garden.sudo.base.game

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

class Cell constructor(var row: Int, var col: Int, var group: Int, var value: Int = 0) {

    var optional = mutableListOf<Int>()
    var valid: Boolean = false
    var preSet: Boolean = false

    @Throws(JSONException::class)
    fun fromJson(jsonObject: JSONObject) {
        val rowT = jsonObject.getInt("row")
        val colT = jsonObject.getInt("col")
        val groupT = jsonObject.getInt("group")
        val valueT = jsonObject.getInt("value")
        val preSetT = jsonObject.getBoolean("preSet")
        val validT = jsonObject.getBoolean("valid")
        val array = jsonObject.getJSONArray("optional")
        var tmp = mutableListOf<Int>()
        for (i in 0 until array.length()){
            tmp.add(array.getInt(i))
        }
        this.row = rowT
        this.col = colT
        this.group = groupT
        this.value = valueT
        this.preSet = preSetT
        this.valid = validT
        this.optional.clear()
        this.optional.addAll(tmp)
    }

    @Throws(JSONException::class)
    fun toJson(): JSONObject {
        val json = JSONObject()
        json.put("row", this.row)
        json.put("col", this.col)
        json.put("group", this.group)
        json.put("valid", this.valid)
        json.put("preSet", this.preSet)
        json.put("value", this.value)
        val array = JSONArray()
        optional.forEach { it ->
            array.put(it)
        }
        json.put("optional", array)
        return json
    }

    override fun toString(): String {
        return "{ row:$row, col:$col, group:$group, value:$value, preSet:$preSet, valid:$valid }"
    }


}